app/assets/javascripts/angular/directives/frame.js.erb

```diff
@@ -7,6 +7,73 @@ angular.module("Prometheus.directives").directive('inlineFrame', function() {
       vars: "="
     },
     restrict: "AE",
-    templateUrl: "<%= asset_path('frame_template.html') %>"
+    templateUrl: "<%= asset_path('frame_template.html') %>",
+    link: function(scope, element, attrs) {
+      // Appended to frame source URL to trigger refresh.
+      scope.refreshCounter = 0;
+      scope.frameHeight = function() {
+        return {
+          height: WidgetHeightCalculator(angular.element(".js_widget_wrapper")[0], scope.aspectRatio)
+        }
+      }
+
+      var a = ['frame.url', 'frame.range', 'frame.endTime', 'refreshCounter'];
+      a.forEach(function(w) {
+        scope.$watch(w, regenFrame)
+      });
+      ['refreshDashboard', 'redrawGraphs'].forEach(function(m) {
+        scope.$on(m, regenFrame);
+      });
+
+      // In order for zoom-out to work correctly on dashboards with iframes,
+      // the iframe needs to be completely reconstructed on each refresh.
+      // Modifying an iframe's src attribute, as in the previous implementation,
+      // changes the browser's history. This interrupts a user's attempt to zoom
+      // out as it takes an additional history.back() execution to traverse the
+      // history state added by modifying an iframe's src.
+      // http://khaidoan.wikidot.com/iframe-and-browser-history
+      function regenFrame() {
+        var url = VariableInterpolator(scope.frame.url, scope.vars);
+        var trustedURL = $sce.trustAsResourceUrl(buildFrameURL(url));
+        var frame = element.find('.frame_container')
+        frame.height(scope.frameHeight().height);
+        frame.html('<iframe src="'+ trustedURL +'" class="frame_iframe" marginwidth="0" scrolling="no"></iframe>');
+      }
+
+      function buildFrameURL(url) {
+        var parser = document.createElement('a');
+        parser.href = url;
+        if (scope.frame.graphite) {
+          return parseGraphiteURL(parser);
+        }
+        parser.search = parser.search + '&decache=' + scope.refreshCounter;
+        return parser.href;
+      }
+
+      function parseGraphiteURL(parser) {
+        var queryStringComponents = parser.search.substring(1).split('&');
+        var fields = {};
+        var targets = [];
+        var FRAME_CONTAINER_PADDING = 10;
```

Comment:
Since this is constant-ish, how about putting it at the top of the file? This way it's even clearer that it's meant to be a constant and no need to rebuild this variable.